### PR TITLE
Added Cell level control for Table Borders

### PR DESCRIPTION
**Pull request by Kaustbh**
Fixes : #1192

Implemented a new feature that allows granular, cell-level control over table borders. With this update, users can now define specific borders (left, right, top, bottom) for individual cells, offering greater customization when generating tables in PDFs.

**Checklist**:

- [ ] The GitHub pipeline is OK (green), 
      meaning that both `pylint` (static code analyzer) and `black` (code formatter) are happy with the changes of this PR.

- [ ] A unit test is covering the code added / modified by this PR

- [ ] This PR is ready to be merged 

- [ ] In case of a new feature, docstrings have been added, with also some documentation in the `docs/` folder

- [ ] A mention of the change is present in `CHANGELOG.md`

By submitting this pull request, I confirm that my contribution is made under the terms of the [GNU LGPL 3.0 license](https://github.com/py-pdf/fpdf2/blob/master/LICENSE).

### CODE REVIEW

**Comment by Kaustbh** [comment]
Hi @Lucas-C, I have made a PR to know if I am doing it correctly it is not a complete implementation, please see the code change and give me suggestions.

File: fpdf/table.py [hunk 1]
```diff
@@ -849,6 +872,7 @@ class Cell:
     rowspan: int
     padding: Optional[Union[int, tuple, type(None)]]
     link: Optional[Union[str, int]]
+    border: Optional[Union[str, int, CellBordersLayout]]
```
Comment (Lucas-C) [inline fpdf/table.py:875]:
```suggestion
    border: Optional[CellBordersLayout]
```

File: fpdf/table.py [hunk 2]
```diff
@@ -819,6 +840,7 @@ def cell(
             rowspan,
             padding,
             link,
+            border,
```
Comment (Lucas-C) [inline fpdf/table.py:843]:
```suggestion
            CellBordersLayout.coerce(border),
```
It's better to check if the value provided is valid early, and to always store an instance of `CellBordersLayout`.

File: fpdf/table.py [hunk 3]
```diff
@@ -251,13 +252,32 @@ def render(self):
...
     def get_cell_border(self, i, j, cell):
         """
         Defines which cell borders should be drawn.
         Returns a string containing some or all of the letters L/R/T/B,
         to be passed to `fpdf.FPDF.multi_cell()`.
         Can be overriden to customize this logic
         """
+
+        if hasattr(cell, "border"):
+            border = CellBordersLayout.coerce(cell.border)
```
Comment (Lucas-C) [inline fpdf/table.py:265]: Not needed if you perform the call to `.coerce()` in `Row.cell()`

File: fpdf/table.py [hunk 4]
```diff
@@ -251,13 +252,32 @@ def render(self):
...
+
+            if border != CellBordersLayout.INHERIT:
+                border2 = []
```
Comment (Lucas-C) [inline fpdf/table.py:268]:
```suggestion
                borders = []
```

**Comment by Lucas-C** [comment]
> Hi @Lucas-C, I have made a PR to know if I am doing it correctly it is not a complete implementation, please see the code change and give me suggestions.

Nice, thank you for creating a PR 🙂

I made a few comments.

On thing that should come next is to add unit tests! 🙂

**Comment by Kaustbh** [comment]
Hi, I made the suggested changes, now how can I fix this pylint warning,
 `fpdf/table.py:255:4: R0911: Too many return statements (7/6) (too-many-return-statements)` 
which is caused because of additional `return str(border)`.

And please tell me where I should add the unit tests.

**Comment by Lucas-C** [comment]
> Hi, I made the suggested changes, now how can I fix this pylint warning, `fpdf/table.py:255:4: R0911: Too many return statements (7/6) (too-many-return-statements)` which is caused because of additional `return str(border)`.

It now seems to be passing, but you have however 27 failing tests.

> And please tell me where I should add the unit tests.

You can add them in `test/table/test_table.py`.

**Comment by Kaustbh** [comment]
Hi @Lucas-C, should I write test just like this? and for all the different values like, `right`, `top`,  etc.

```
def test_cell_border_left(tmp_path):
    pdf = FPDF()
    pdf.add_page()
    pdf.set_font("Times", size=5)
    with pdf.table( gutter_height=2, gutter_width=2) as table:

        for data_row in TABLE_DATA:
            row = table.row()
            for datum in data_row:
                row.cell(datum,border="left")
    assert_pdf_equal(pdf, HERE / "test_cell_border_left.pdf", tmp_path)
```

**Comment by Kaustbh** [comment]
@Lucas-C is it the right way of writing the tests?

**Comment by Lucas-C** [comment]
> Hi @Lucas-C, should I write test just like this? and for all the different values like, `right`, `top`, etc.

Yes, this look likes a good start 👍
As mentioned on https://py-pdf.github.io/fpdf2/Development.html, you can pass `generate=True` once to `assert_pdf_equal()` in order to generate a reference PDF file.

And testing many possible combination of values is also a good idea!

**Comment by Kaustbh** [comment]
is it fine @Lucas-C ?

File: test/text/test_cell.py [hunk 5]
```diff
@@ -362,3 +361,239 @@ def test_cell_speed_with_long_text():  # issue #907
     assert len(LONG_TEXT_LINES) == 26862
     for line in LONG_TEXT_LINES:
         pdf.cell(0, 3, line, new_x="LMARGIN", new_y="NEXT")
+
+
+def test_cell_border_none(tmp_path):
```
Comment (Lucas-C) [inline test/text/test_cell.py:366]: Could you regroup all or most of those tests in a single one, with one case per page, in order to limit the number of reference PDF files, please?

All the right/left/top/bottom cases could form a single unit test I think.

`inherit` should have its own dedicated unit test & reference file, but I think more cases should be tested:
* cell `border="INHERIT"` with various table `borders_layout` values
* cell `border="INHERIT"` with table `outer_border_width` set

File: fpdf/table.py (hunk 3 above)
Comment (Lucas-C) [reply fpdf/table.py:265]: I do not think that those `hasattr` & `CellBordersLayout.coerce` calls are needed.

**Review by Lucas-C — CHANGES_REQUESTED** [review]
> is it fine @Lucas-C ?

There are a couple of minor things remaining to be fixed, and also could you please add a mention of this addition in the `CHANGELOG.md` file?

The I'll be happy to merge your PR 👍 🙂

**Comment by Kaustbh** [comment]
Hi @Lucas-C,

Should I add documentation for this as well, or would it be better to address it in a new issue? Creating a separate issue might also be a good idea, as it would give me the opportunity to make an additional contribution 🙂

**Comment by Lucas-C** [comment]
> Should I add documentation for this as well, or would it be better to address it in a new issue? Creating a separate issue might also be a good idea, as it would give me the opportunity to make an additional contribution 🙂

We usually integrate the documentation related to new features as part of the PR that introduced it, so yes, it would be a good idea to include some documentation as part of this PR 🙂

Also, have you seen my other previous comments regarding the remaing minor changes required?

**Comment by Kaustbh** [comment]
Yes, I have made the changes you suggested.

**Comment by Lucas-C** [comment]
> Yes, I have made the changes you suggested.

I'm sorry but I don't think you adressed all those points:
* https://github.com/py-pdf/fpdf2/pull/1285#discussion_r1821995647
* https://github.com/py-pdf/fpdf2/pull/1285#discussion_r1821993474

And I also previously wrote:

> also could you please add a mention of this addition in the CHANGELOG.md file?

> We usually integrate the documentation related to new features as part of the PR that introduced it, so yes, it would be a good idea to include some documentation as part of this PR 🙂

**Comment by Kaustbh** [comment]
I haven't committed it yet.

**Comment by Lucas-C** [comment]
> I haven't committed it yet.

Oh OK, alright, that's fine🙂
Take your time 👍

**Comment by Kaustbh** [comment]
Hi @Lucas-C, I have made the changes and also added documentation. I'm sorry, but I was not able to render the API documentation on my laptop. Please review the code and let me know if any changes are required.

File: CHANGELOG.md [hunk 6]
```diff
@@ -18,6 +18,8 @@ This can also be enabled programmatically with `warnings.simplefilter('default',
 
 ## [2.8.2] - Not released yet
 ### Added
+* new optional parameter `border` for table cells [issue #1192](https://github.com/py-pdf/fpdf2/issues/1192) users can define specific borders (left, right, top, bottom) for individual cells
+* support for overriding paragraph direction on bidirectional text
```
Comment (Lucas-C) [inline CHANGELOG.md:22]:
```suggestion
```
This line comes from release `2.7.9`

File: docs/Tables.md [hunk 7]
```diff
@@ -277,6 +278,61 @@ Result:
...
+
+```python
+pdf = FPDF()
+pdf.add_page()
+pdf.set_font("Times", size=16)
+with pdf.table() as table:
+    for data_row in TABLE_DATA:
+        row = table.row()
+        for datum in data_row:
+            row.cell(datum, border=1) # Left
```
Comment (Lucas-C) [inline docs/Tables.md:293]:
```suggestion
            row.cell(datum, border="LEFT")
```

File: docs/Tables.md [hunk 8]
```diff
@@ -277,6 +278,61 @@ Result:
...
+
+```python
+pdf = FPDF()
+pdf.add_page()
+pdf.set_font("Times", size=16)
+with pdf.table() as table:
+    for data_row in TABLE_DATA:
+        row = table.row()
+        for datum in data_row:
+            row.cell(datum, border=4) # Top
```
Comment (Lucas-C) [inline docs/Tables.md:309]:
```suggestion
            row.cell(datum, border="TOP")
```

File: docs/Tables.md [hunk 9]
```diff
@@ -277,6 +278,61 @@ Result:
 All the possible layout values are described
 there: [`TableBordersLayout`](https://py-pdf.github.io/fpdf2/fpdf/enums.html#fpdf.enums.TableBordersLayout).
 
+## Set cell borders
+
+_New in [:octicons-tag-24: 2.8.2](https://github.com/py-pdf/fpdf2/blob/master/CHANGELOG.md)_
+
+```python
+pdf = FPDF()
```
Comment (Lucas-C) [inline docs/Tables.md:286]:
```suggestion
from fpdf import FPDF

pdf = FPDF()
```
We try to make all code snippets autonomous: you can copy paste it in a `.py` script and it should just work!
That's why including the imports is important 🙂

File: docs/Tables.md [hunk 10]
```diff
@@ -277,6 +278,61 @@ Result:
...
+            row.cell(datum, border=1) # Left 
+pdf.output('table.pdf')
+```
+
+Result:
+
+![](table_with_cell_border_left.jpg)
+
+```python
+pdf = FPDF()
```
Comment (Lucas-C) [inline docs/Tables.md:302]:
```suggestion
from fpdf import FPDF

pdf = FPDF()
```
We try to make all code snippets autonomous: you can copy paste it in a `.py` script and it should just work!
That's why including the imports is important 🙂

File: docs/Tables.md [hunk 11]
```diff
@@ -277,6 +278,61 @@ Result:
...
+            row.cell(datum, border=4) # Top
+pdf.output('table.pdf')
+```
+
+Result:
+
+![](table_with_cell_border_top.jpg)
+
+```python
+pdf = FPDF()
```
Comment (Lucas-C) [inline docs/Tables.md:318]:
```suggestion
from fpdf import FPDF

pdf = FPDF()
```
We try to make all code snippets autonomous: you can copy paste it in a `.py` script and it should just work!
That's why including the imports is important 🙂

File: docs/Tables.md [hunk 12]
```diff
@@ -277,6 +278,61 @@ Result:
...
+pdf.add_page()
+pdf.set_font("Times", size=16)
+with pdf.table() as table:
+    for data_row in TABLE_DATA:
+        row = table.row()
+        for datum in data_row:
+            row.cell(datum, border=5) # Left + Top
```
Comment (Lucas-C) [inline docs/Tables.md:325]:
```suggestion
            row.cell(datum, border=CellBordersLayout.TOP | CellBordersLayout.LEFT)
```

File: fpdf/table.py (hunk 3 above)
Comment (Lucas-C) [reply fpdf/table.py:265]: I tested on your branch, an lines 264 to 268 can be replaced by:
```python
        if cell.border != CellBordersLayout.INHERIT:
            return str(cell.border)
```

**Review by Lucas-C — APPROVED** [review]
Good job overall 👍 👍 👍

You wrote really good code, good documentation and good tests!

There are only a few minor details remaining:
* this early comment I made that I think you might have missed: https://github.com/py-pdf/fpdf2/pull/1285#discussion_r1831595402
* an extra line in `CHANGELOG.md`: https://github.com/py-pdf/fpdf2/pull/1285#discussion_r1831582944
* some minor improvements in `docs/Tables.md`

**Comment by Kaustbh** [comment]
@Lucas-C  please review the commit.

**Review by Lucas-C — APPROVED** [review]

**Comment by Lucas-C** [comment]
@allcontributors pleasae add @Kaustbh for code

**Comment by Lucas-C** [comment]
Merged!

Thanks a lot for your contribution @Kaustbh 👍

You have been added to the list of `fpdf2` contributors: https://github.com/py-pdf/fpdf2?tab=readme-ov-file#contributors-

**Comment by Kaustbh** [comment]
It was a pleasure contributing and I'm honored to be included in the list of contributors. I am looking forward to collaborating more in the future and contributing to more features and improvements.

**Pull request merged**